### ISSUE TITLE
Store member door code in the main DU app

### ISSUE THREAD

**Issue by compwron** [doubleunion/arooo#360]
Where in the app would we store this? I would say on the `User` object, or possibly add a new database table door_codes which has columns: `id, user_id, door_code, created_at, updated_at`
We should *show* the door code (masked?) on a users private view of their own page, and add a way for an admin to add a door code to a user

Making a member a past member should deactivate their door code. 

**Benefits:**
- less code to maintain
- door codes in a database instead of in a codebase which we must keep secret
- makes it easier to update the code that we use to talk to Twilio, since it will be public and part of the main codebase
- it will be much easier for DU admins to add/update/remove door codes (git knowledge will no longer be required)

**Concerns:**
- We need to be careful with the migration process, we might lock everyone out if we mess up
- Will this make the door codes less secure? (In my opinion no- the small group of people who have heroku / rails console access is very similar or possibly identical who the people who have access to the doorcodes code right now)
- Should we encrypt the door codes in the database? On the one hand, we need to decrypt them to interact with Twilio anyway, and anyone who can interact with the app console and database would be able to decrypt the door codes anyway. On the other hand... this is the kind of thing that should be kept secure. 
- After a member stops being a member, their doorcode should stop working. It should not be usable by new members, because there's no way to be sure that the old member won't just use the doorcode again since they know it :) (does this even matter? we mostly don't worry about this kind of thing... but security is worth doing)

**Steps**

- Add a new column to the User table. It must be unique. Assume that if a member stops being a member, their code will not be available for reuse (until we run out of codes, but that is a long long time away)
- Load the existing member data into the table
- Add a UI field for each member to see their own key code. Do not yet allow admins to change or set them.
- Copy over the twilio logic and codes from the du-doorbell app (but don't start using it yet)
- Physically go to the DU space, turn off the du-doorbell app, turn on the arooo doorbell code, check that it works

**Comment by carmiendo** [comment]
Another nice benefit of this is that we could make the code for each person
visible on their app page, so they can easily look it up if they forget it!

**Comment by compwron** [comment]
Guide to the current process of adding/deleting from the door app: https://docs.google.com/document/d/1-GbnJz8RE5wS7iz0RtYTZe8H3dQqt2kukgGfs0n-dLg/edit (internal access only)

**Comment by compwron** [comment]
member: https://doubleunion.slack.com/archives/CKDA9B9T4/p1569704450000700
blergh I’m running way late, had to re-look-up my keycode and can’t find it so I’m just going to come over

**Comment by compwron** [comment]
We should definitely do this. 

Steps:
Reach out to du-doorbell maintainers to get buy-in and advice (check git history)
Add a column to the member table with a key code
Load the existing member data into 
Add a UI field for each member to see their own key code
Copy over the twilio logic and codes from the du-doorbell app (but don't start using it yet)
Physically go to the DU space, turn off the du-doorbell app, turn on the arooo doorbell code, check that it works
Rejoice

**Comment by anaulin** [comment]
@compwron and myself just fleshed out the design and steps to get this done in this doc: https://docs.google.com/document/d/1utd6l6lxcLpFfEpg5FBeu_xTXJ66bS1C9fjQ831fuIE/edit?usp=sharing
(link should be visible to DU members)

**Comment by anaulin** [comment]
@compwron and anyone else following along: I'm working on adding the `doorcodes` table, as the first step in this migration, and even with the simplest migration (a table with just `t.timestamps`), I end up with a bunch of errors being written to `schema.rb`, all of them like this:
```ruby
# Could not dump table "profiles" because of following FrozenError
#   can't modify frozen String: "false"
[...]
# Could not dump table "applications" because of following FrozenError
#   can't modify frozen String: "false"
[...]
# Could not dump table "users" because of following FrozenError
#   can't modify frozen String: "false"
[...]
````

Sadly, the errors don't tell me where the heck is this frozen "false" string, so hard to debug. 🙄 

I tried nuking my local database and recreating everything from scratch, in case it was my local db was in a bad state. It didn't help.

Searching around, the best I could find is that this is an issue related to incompatibilities between Ruby 2.7.1 (to which we've upgraded since the last migration was run), and Rails 4.2.*. Here is another person having the same issue with this version combination, and finding only advice to either upgrade Rails or downgrade Ruby: https://stackoverflow.com/questions/62570662/rails-4-ruby-2-7-1-schema-rb-shows-could-not-dump-table-because-of-following-f

According to the compatibility table at https://www.fastruby.io/blog/ruby/rails/versions/compatibility-table.html, we should be using Ruby 2.2 with Rails 4.2. 

*Long story short, unless someone has better insight into this problem, we're going to either upgrade Rails (which I think would be preferable, since we're way behind), or downgrade our Ruby version (which would be easier, but sadness).*

**Comment by anaulin** [comment]
Good news. Looks like I no longer have this migration issue, once I rebased my migration onto the upgraded Rails from PR https://github.com/doubleunion/arooo/pull/441

**Comment by anaulin** [comment]
I have imported most of the existing door codes into the Arooo database, and set to `enabled` the ones that are enabled in the `du-doorbell` app. For future reference and repeatability, I checked into `du-doorbell` a file, named `import_codes_users.rb` that has the data, code and instructions that I used to do this import.

The file `import_codes_users.rb` also has a commented-out list of the codes that I could not import because I cannot figure out which users own them. (E.g. there are multiple codes that are owned by "Liz"). We should figure out what do do about these codes at some point, but this doesn't seem to be urgent right now, given that access to the space is very restricted.

I believe this import puts us at the point where we can switch over Twilio webhooks to call into the migrated doorbell code in Arooo, and deprecate the `du-doorbell` app (but let's keep it around for a while, until we are certain we don't need to switch back).

I'll start a convo in the Slack #du-app channel to coordinate on the precise timing and logistics of doing the switchover of Twilio to Arooo.

fyi @Imperiopolis @compwron

**Comment by Imperiopolis** [comment]
This is awesome, thanks @anaulin. Please let me know via email or DM when things are cut over or if you need help switching over / testing the new system.

**Comment by anaulin** [comment]
I just did the switchover of the Twilio configuration to call Arooo, and verified that it works with some manual tests.

I propose calling this Issue "done", and opening separate issues for the "nice to have" follow-up tasks that are left in the planning doc (https://docs.google.com/document/d/1utd6l6lxcLpFfEpg5FBeu_xTXJ66bS1C9fjQ831fuIE/edit).

**Comment by anaulin** [comment]
Most of the items in the todo list at https://docs.google.com/document/d/1utd6l6lxcLpFfEpg5FBeu_xTXJ66bS1C9fjQ831fuIE/edit#heading=h.o2tf67a7dotj are now complete.

Two remaining items have their own issues (https://github.com/doubleunion/arooo/issues/477 and https://github.com/doubleunion/arooo/issues/478), and the third and final remaining to-do is "turn off the doorbell app in Heroku", which I think we should wait some months to do.
